protocol/bin-prot/src/de.rs

```diff
@@ -344,6 +355,54 @@ impl<'de, 'a, R: Read> de::Deserializer<'de> for &'a mut Deserializer<R, Strongl
     }
 }
 
+pub struct PolyvarEnum<'a, R: Read, Mode> {
+    de: &'a mut Deserializer<R, Mode>,
+    hash: VariantHash,
+    variants: &'a [&'static str],
+}
+
+impl<'a, 'de, R: Read, Mode> PolyvarEnum<'a, R, Mode> {
+    pub fn new(
+        de: &'a mut Deserializer<R, Mode>,
+        hash: VariantHash,
+        variants: &'a [&'static str],
+    ) -> Self {
+        PolyvarEnum { de, hash, variants }
+    }
+}
+
+macro_rules! impl_polyvar_enum_access {
+    ($ty: ty) => {
+        impl<'de, 'a, R: Read> EnumAccess<'de> for PolyvarEnum<'a, R, $ty> {
+            type Error = Error;
+            type Variant = Enum<'a, R, StronglyTyped>;
+
+            fn variant_seed<V>(self, seed: V) -> Result<(V::Value, Self::Variant)>
+            where
+                V: de::DeserializeSeed<'de>,
+            {
+                let mut index: Option<usize> = None;
+                for (i, v) in self.variants.into_iter().enumerate() {
+                    // Return the first polyvar variant where the hash matches
+                    if caml_hash_variant(v)? == self.hash {
+                        index = Some(i);
+                    }
+                }
+                match index {
+                    Some(i) => {
+                        let de: U8Deserializer<Self::Error> = (i as u8).into_deserializer();
+                        let v = seed.deserialize(de)?;
+                        Ok((v, Enum::new(self.de, i as u8)))
+                    }
+                    None => Err(Error::UnknownPolyvarTag(self.hash)),
```

Comment:
@willemolding since we cannot use `caml_hash_variant?` inside `find_map`, I had to change the previous code to this

we could have used something like 
```
let index = self
       .variants
       .into_iter()
       .enumerate()
       .find_map(|(index, value)| {
           if caml_hash_variant(value).unwrap() == self.hash {
                 Some(index)
            } else {
                 None
            }
       })
       .ok_or(Error::UnknownPolyvarTag(self.hash))?;
```
but using unwrap/expect inside a lib is not a good idea

Reply:
You could expand out the `?` operator and it works quite nicely

```rust
                let index = self
                    .variants
                    .into_iter()
                    .enumerate()
                    .find_map(|(index, value)| {
                        match caml_hash_variant(value) {
                            Ok(hash) => {
                                if hash == self.hash {
                                    Some(Ok(index))
                                } else {
                                    None
                                }
                            }
                            Err(e) => Some(Err(e))
                        }
                    })
                    .ok_or(Error::UnknownPolyvarTag(self.hash))??;
```

Reply:
thanks, this is perfect

Reply:
Hey but won't this return early because we get an error being returned from find_map which is Some(err)?

Reply:
yeah this will throw an error as soon as a non ascii polyvar tag is encountered

Reply:
@creativcoder @willemolding code suggestion
any suggestion for making it cleaner
```
                let mut index = None;
                let mut error = None;
                for (i, v) in self.variants.into_iter().enumerate() {
                    // Return the first polyvar variant where the hash matches
                    match caml_hash_variant(v) {
                        Ok(val) => {
                            if val == self.hash {
                                index = Some(i)
                            }
                        }
                        Err(err) => {
                            error = Some(err);
                        }
                    }
                }
                match index {
                    Some(i) => {
                        let de: U8Deserializer<Self::Error> = (i as u8).into_deserializer();
                        let v = seed.deserialize(de)?;
                        Ok((v, Enum::new(self.de, i as u8)))
                    }
                    None => match error {
                        Some(err) => Err(err),
                        None => Err(Error::UnknownPolyvarTag(self.hash)),
                    },
                }
```

Reply:
or we could take out ascii check from ocaml_hash_variant method and do 
```
               let mut error = None;
                // Return the first polyvar variant where the hash matches
                let index = self
                    .variants
                    .into_iter()
                    .enumerate()
                    .find_map(|(index, value)| {
                        if !value.is_ascii() {
                            error = Some(Error::PolyvarTagNotAscii);
                            None
                        } else if caml_hash_variant(value) == self.hash {
                            Some(index)
                        } else {
                            None
                        }
                    })
                    .ok_or(Error::UnknownPolyvarTag(self.hash))?;
                let de: U8Deserializer<Self::Error> = (index as u8).into_deserializer();
                let v = seed.deserialize(de)?;
                Ok((v, Enum::new(self.de, index as u8)))
```

Reply:
The second looks clean, but where do we use the `error` after we set it to Some(err)